### ISSUE TITLE
Docker image reported by status is not the image digest

### ISSUE THREAD

**Issue by smarterclayton** [kubernetes/kubernetes#14047]
The image digest from an image is much more useful to be reported with a container than just the image id.  We should return the digest as well, which would allow policy tie back to 3rd party engines as well as understanding the exact image in use.  I believe the gap today is the local graph driver doesn't preserve the digest, but that work is going on upstream in Docker to make that more generally available.

@vbatts correct me if I missed anything about the state of Docker on this (links to issues would be helpful)

@ncdc @pweil- policy around managing what images can be used on a Kubelet is a next big area of exploration - knowing exactly what is in use is a part of that

**Comment by smarterclayton** [comment]
@kubernetes/rh-platform-management

**Comment by ncdc** [comment]
When you pull by tag, the graph driver records the association between the tag and the v1 image ID. If you are displaying the pull progress, you'll see something like this:

```
$ docker pull busybox
Using default tag: latest
Trying to pull repository docker.io/library/busybox ... latest: Pulling from library/busybox
cf2616975b4a: Already exists
6ce2e90b0bc7: Already exists
8c2e06607696: Already exists
Digest: sha256:9fbfab5715af5aba7322931ec1f03c6b3978109b68053f4b0ba1d97abb93e21c
Status: Image is up to date for docker.io/busybox:latest
```

As you can see, the digest is displayed (`Digest: sha256:9fbfab5715af5aba7322931ec1f03c6b3978109b68053f4b0ba1d97abb93e21c`) but this is purely informational; the graph driver doesn't record the digest, and it is "lost" from the daemon's perspective.

If you want to be able to display the digest, you'll have to pull by digest and reference the image by digest as well.

Going forward, I would hope that the graph driver will be updated to store the downloaded image manifest, which would include the digest, and this would no longer be a gap.

**Comment by smarterclayton** [comment]
Closing this is pretty important for audit / policy.  Let's try to use this
as a tracking issue.

On Wed, Sep 16, 2015 at 1:21 PM, Andy Goldstein <notifications@github.com>
wrote:

> When you pull by tag, the graph driver records the association between the
> tag and the v1 image ID. If you are displaying the pull progress, you'll
> see something like this:
>
> $ docker pull busybox
> Using default tag: latest
> Trying to pull repository docker.io/library/busybox ... latest: Pulling from library/busybox
> cf2616975b4a: Already exists
> 6ce2e90b0bc7: Already exists
> 8c2e06607696: Already exists
> Digest: sha256:9fbfab5715af5aba7322931ec1f03c6b3978109b68053f4b0ba1d97abb93e21c
> Status: Image is up to date for docker.io/busybox:latest
>
> As you can see, the digest is displayed (Digest:
> sha256:9fbfab5715af5aba7322931ec1f03c6b3978109b68053f4b0ba1d97abb93e21c)
> but this is purely informational; the graph driver doesn't record the
> digest, and it is "lost" from the daemon's perspective.
>
> If you want to be able to display the digest, you'll have to pull by
> digest and reference the image by digest as well.
>
> Going forward, I would hope that the graph driver will be updated to store
> the downloaded image manifest, which would include the digest, and this
> would no longer be a gap.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/issues/14047#issuecomment-140810622>
> .
>



-- 
Clayton Coleman | Lead Engineer, OpenShift

**Comment by ncdc** [comment]
1 possible option would be to have Kubernetes download the manifest for the tag by talking to the registry directly. The digest is included in a response header. We could also see if the docker/distribution registry client can give us that info, which would be nicer.

**Comment by dustymabe** [comment]
> 1 possible option would be to have Kubernetes download the manifest for the tag by talking to the registry directly. The digest is included in a response header. We could also see if the docker/distribution registry client can give us that info, which would be nicer.

Does `skopeo` (or really the underlying libraries) give us this? 

```
[root@f24 ~]# skopeo inspect docker://docker.io/fedora:24                                                                                              
{
    "Name": "docker.io/library/fedora",
    "Digest": "sha256:64a02df6aac27d1200c2572fe4b9949f1970d05f74d367ce4af994ba5dc3669e",
    "RepoTags": [
        "20",
        "21",
        "22",
        "23",
        "24",
        "heisenbug",
        "latest",
        "rawhide"
    ],
    "Created": "2016-08-15T21:46:15.78392612Z",
    "DockerVersion": "1.10.3",
    "Labels": {},
    "Architecture": "amd64",
    "Os": "linux",
    "Layers": [
        "sha256:2bf01635e2a0f7ed3800c8cb3effc5ff46adc6b9b86f0e80743c956371efe553"
    ]
}

```

**Comment by ncdc** [comment]
@dustymabe this is getting fixed by #33014

**Comment by dustymabe** [comment]
> @dustymabe this is getting fixed by #33014

Thanks! Just as a side question.. does this mean a user would be able to specify an image by "digest" rather than just name:tag when creating new pods within kubernetes?

**Comment by ncdc** [comment]
@dustymabe that has been possible since docker 1.6 :-)

**Comment by dustymabe** [comment]
:) - I was trying to use the digest references with an older version of k8s/openshift - I moved to a newer version and it works

**Comment by yujuhong** [comment]
Fixed by #33014. Closing the issue. We should revisit *what* should be exposed in the pod status in a separate issue:
 * Image config digest
 * Arbitrary image ID generated by the container runtime
 * A (normalized) pullable reference such as RepoDigest or RepoTags. 
 * ...
